File: benchmarks/go/src/mt_mergesort.go
package main

import (
	"flag"
	"fmt"
	"os"
	"sync"
	"unsafe"
)

// ThreadArgs holds the arguments for a merge sort operation
type ThreadArgs struct {
	arr       []int32
	left      int
	right     int
	depth     int
	maxDepth  int
}

// merge combines two sorted subarrays into a single sorted array
func merge(arr []int32, left, mid, right int) {
	n1 := mid - left + 1
	n2 := right - mid

	// Create temporary arrays
	L := make([]int32, n1)
	R := make([]int32, n2)

	// Copy data to temporary arrays
	copy(L, arr[left:left+n1])
	copy(R, arr[mid+1:mid+1+n2])

	// Merge the temporary arrays back
	i, j, k := 0, 0, left
	for i < n1 && j < n2 {
		if L[i] <= R[j] {
			arr[k] = L[i]
			i++
		} else {
			arr[k] = R[j]
			j++
		}
		k++
	}

	// Copy remaining elements of L[]
	for i < n1 {
		arr[k] = L[i]
		i++
		k++
	}

	// Copy remaining elements of R[]
	for j < n2 {
		arr[k] = R[j]
		j++
		k++
	}
}

// mergeSortThread performs parallel merge sort using goroutines
func mergeSortThread(args *ThreadArgs) {
	arr := args.arr
	left := args.left
	right := args.right
	depth := args.depth
	maxDepth := args.maxDepth

	if left < right {
		mid := left + (right-left)/2

		if depth < maxDepth {
			// Create goroutines for left and right halves
			var wg sync.WaitGroup
			wg.Add(2)

			// Process left half
			go func() {
				defer wg.Done()
				leftArgs := &ThreadArgs{
					arr:      arr,
					left:     left,
					right:    mid,
					depth:    depth + 1,
					maxDepth: maxDepth,
				}
				mergeSortThread(leftArgs)
			}()

			// Process right half
			go func() {
				defer wg.Done()
				rightArgs := &ThreadArgs{
					arr:      arr,
					left:     mid + 1,
					right:    right,
					depth:    depth + 1,
					maxDepth: maxDepth,
				}
				mergeSortThread(rightArgs)
			}()

			wg.Wait()
		} else {
			// Sequential sorting for remaining depth
			leftArgs := &ThreadArgs{
				arr:      arr,
				left:     left,
				right:    mid,
				depth:    depth + 1,
				maxDepth: maxDepth,
			}
			rightArgs := &ThreadArgs{
				arr:      arr,
				left:     mid + 1,
				right:    right,
				depth:    depth + 1,
				maxDepth: maxDepth,
			}
			mergeSortThread(leftArgs)
			mergeSortThread(rightArgs)
		}

		merge(arr, left, mid, right)
	}
}

func main() {
	// Parse command line arguments
	flag.Parse()
	args := flag.Args()
	if len(args) != 4 {
		fmt.Fprintf(os.Stderr, "Usage: %s <input_file> <num_integers> <num_cores> <output_file>\n", os.Args[0])
		os.Exit(1)
	}

	inputFile := args[0]
	numIntegers := parseInt(args[1])
	numCores := parseInt(args[2])
	outputFile := args[3]

	// Calculate max depth for goroutine creation
	maxDepth := 0
	temp := numCores
	for temp > 1 {
		maxDepth++
		temp /= 2
	}

	// Allocate memory for the array
	arr := make([]int32, numIntegers)

	// Read input file
	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening input file: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	// Read all integers at once
	_, err = file.Read(unsafe.Slice((*byte)(unsafe.Pointer(&arr[0])), numIntegers*4))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error reading input file: %v\n", err)
		os.Exit(1)
	}

	// Perform parallel merge sort
	threadArgs := &ThreadArgs{
		arr:      arr,
		left:     0,
		right:    len(arr) - 1,
		depth:    0,
		maxDepth: maxDepth,
	}
	mergeSortThread(threadArgs)

	// Write output file
	outFile, err := os.Create(outputFile)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error creating output file: %v\n", err)
		os.Exit(1)
	}
	defer outFile.Close()

	// Write all integers at once
	_, err = outFile.Write(unsafe.Slice((*byte)(unsafe.Pointer(&arr[0])), len(arr)*4))
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error writing output file: %v\n", err)
		os.Exit(1)
	}
}

func parseInt(s string) int {
	var i int
	_, err := fmt.Sscanf(s, "%d", &i)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error parsing integer: %v\n", err)
		os.Exit(1)
	}
	return i
}

File: benchmarks/nodejs/mt-mergesort.js
const fs = require('fs');
const { Worker, isMainThread, parentPort, workerData } = require('worker_threads');

function merge(arr, left, mid, right) {
    const n1 = mid - left + 1;
    const n2 = right - mid;

    // Create temporary arrays
    const L = new Int32Array(n1);
    const R = new Int32Array(n2);

    // Copy data to temporary arrays using memcpy-like approach
    L.set(arr.slice(left, left + n1));
    R.set(arr.slice(mid + 1, mid + 1 + n2));

    // Merge the temporary arrays back
    let i = 0, j = 0, k = left;
    while (i < n1 && j < n2) {
        if (L[i] <= R[j]) {
            arr[k] = L[i];
            i++;
        } else {
            arr[k] = R[j];
            j++;
        }
        k++;
    }

    // Copy remaining elements of L[]
    while (i < n1) {
        arr[k] = L[i];
        i++;
        k++;
    }

    // Copy remaining elements of R[]
    while (j < n2) {
        arr[k] = R[j];
        j++;
        k++;
    }
}

async function mergeSortThread(arr, left, right, depth, maxDepth) {
    if (left < right) {
        const mid = Math.floor(left + (right - left) / 2);

        if (depth < maxDepth) {
            // Create promises for parallel execution
            const leftPromise = new Promise((resolve) => {
                const worker = new Worker(__filename, {
                    workerData: {
                        arr: arr.slice(left, mid + 1),
                        left: 0,
                        right: mid - left,
                        depth: depth + 1,
                        maxDepth: maxDepth
                    }
                });
                worker.on('message', (sortedLeft) => {
                    arr.set(sortedLeft, left);
                    resolve();
                });
            });

            const rightPromise = new Promise((resolve) => {
                const worker = new Worker(__filename, {
                    workerData: {
                        arr: arr.slice(mid + 1, right + 1),
                        left: 0,
                        right: right - (mid + 1),
                        depth: depth + 1,
                        maxDepth: maxDepth
                    }
                });
                worker.on('message', (sortedRight) => {
                    arr.set(sortedRight, mid + 1);
                    resolve();
                });
            });

            await Promise.all([leftPromise, rightPromise]);
        } else {
            // Sequential sorting for remaining depth
            await mergeSortThread(arr, left, mid, depth + 1, maxDepth);
            await mergeSortThread(arr, mid + 1, right, depth + 1, maxDepth);
        }

        merge(arr, left, mid, right);
    }
}

// Worker thread code
if (!isMainThread) {
    const { arr, left, right, depth, maxDepth } = workerData;
    const workerArr = new Int32Array(arr);
    mergeSortThread(workerArr, left, right, depth, maxDepth)
        .then(() => parentPort.postMessage(workerArr));
}

// Main thread code
if (isMainThread) {
    if (process.argv.length !== 6) {
        console.error(`Usage: ${process.argv[1]} <input_file> <num_integers> <num_cores> <output_file>`);
        process.exit(1);
    }

    const inputFile = process.argv[2];
    const numIntegers = parseInt(process.argv[3]);
    const numCores = parseInt(process.argv[4]);
    const outputFile = process.argv[5];

    // Calculate max depth for thread creation
    let maxDepth = 0;
    let temp = numCores;
    while (temp > 1) {
        maxDepth++;
        temp = Math.floor(temp / 2);
    }

    // Read input file
    let buffer;
    try {
        buffer = fs.readFileSync(inputFile);
    } catch (err) {
        console.error('Error opening input file');
        process.exit(1);
    }

    const arr = new Int32Array(buffer.buffer, buffer.byteOffset, numIntegers);

    // Perform parallel merge sort
    mergeSortThread(arr, 0, numIntegers - 1, 0, maxDepth)
        .then(() => {
            try {
                const outputBuffer = Buffer.from(arr.buffer, arr.byteOffset, numIntegers * 4);
                fs.writeFileSync(outputFile, outputBuffer);
            } catch (err) {
                console.error('Error writing output file');
                process.exit(1);
            }
        })
        .catch(err => {
            console.error('Error during merge sort:', err);
            process.exit(1);
        });
}
